### PR TITLE
Run raspberry pi tests.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -75,6 +75,7 @@ env:
   DO_RELEASE: ${{ github.event_name == 'release' || startsWith(github.event.inputs.do-release, 'v') }}
   DO_MORE_TESTS: ${{ github.event_name == 'schedule' || github.event.inputs.test-more == 'true' }}
   ESPTOOL_VERSION: "v4.8.1%2Btoitlang"
+  TEST_PI_HW: "true"
```
Comment (kasperl) [inline .github/workflows/ci.yml:78]: Somehow, I like grouping the boolean flags that decide on how much to run, build, test under the `DO_` prefix.

```suggestion
  DO_TEST_PI_HW: "true"
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1221,3 +1271,37 @@ jobs:
...
+          rm -rf ${{ github.workspace }}/*
+
+      - uses: actions/checkout@v4
+
+      # We need a more recent version of cmake than is installed globally.
+      - name: Add local bin to PATH
+        run: |
+          echo "$HOME/local/bin" >> $GITHUB_PATH
+
+      # Download the pi artifact.
```
Comment (kasperl) [inline .github/workflows/ci.yml:1293]:
```suggestion
      # Download the Pi artifact.
```

File: Makefile [hunk 3]
```diff
@@ -306,6 +306,17 @@ install: install-sdk
 test:
 	(cd $(BUILD)/$(HOST) && ninja check_slow check_fuzzer_lib)
 
+.PHONY: test-serial
+test-serial:
+	@if [ -z "$$HW_TOIT_EXE" ]; then \
+		echo "HW_TOIT_EXE is not set. Set it to the path of the PI executable."; \
```
Comment (kasperl) [inline Makefile:312]:
```suggestion
		echo "HW_TOIT_EXE is not set. Set it to the path of the Pi executable."; \
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1221,3 +1271,37 @@ jobs:
...
+      - name: Unpack Raspberry Pi
+        run: |
+          mkdir -p rpi
+          tar x -zf rpi.tar.gz -C rpi
+
+      - name: Test Raspberry Pi
+        run: |
+          # Get the pin configuration from the file in the runner's home.
+          source $HOME/pi-test.env
+          make HW_TOIT_EXE=${{ github.workspace }}/rpi/toit/bin/toit test-serial
```
Comment (kasperl) [inline .github/workflows/ci.yml:1307]:
```suggestion
          make TOIT_EXE_HW=${{ github.workspace }}/rpi/toit/bin/toit test-serial
```

File: Makefile [hunk 5]
```diff
@@ -306,6 +306,17 @@ install: install-sdk
 test:
 	(cd $(BUILD)/$(HOST) && ninja check_slow check_fuzzer_lib)
 
+.PHONY: test-serial
+test-serial:
+	@if [ -z "$$HW_TOIT_EXE" ]; then \
```
Comment (kasperl) [inline Makefile:311]:
```suggestion
	@if [ -z "$$TOIT_EXE_HW" ]; then \
```

File: Makefile [hunk 6]
```diff
@@ -306,6 +306,17 @@ install: install-sdk
...
+		echo "HW_TOIT_EXE is not set. Set it to the path of the PI executable."; \
+		exit 1; \
+	fi
+	mkdir -p $(BUILD)/serial
+	(cd $(BUILD)/serial && cmake -DHW_TOIT_EXE=$$HW_TOIT_EXE -G Ninja $(CURDIR)/tests/hw)
```
Comment (kasperl) [inline Makefile:316]:
```suggestion
	(cd $(BUILD)/serial && cmake -DTOIT_EXE_HW=$$TOIT_EXE_HW -G Ninja $(CURDIR)/tests/hw)
```

File: tests/hw/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# Lesser General Public License for more details.
+#
+# The license can be found in the file `LICENSE` in the top level
+# directory of this repository.
+
+cmake_minimum_required(VERSION 3.20.0)
+
+project(hw NONE)  # No need for compilers.
+
+set(HW_TOIT_EXE "" CACHE FILEPATH "The executable used to run hw tests")
```
Comment (kasperl) [inline tests/hw/CMakeLists.txt:20]:
```suggestion
set(TOIT_EXE_HW "" CACHE FILEPATH "The executable used to run hardware tests")
```

File: tests/hw/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+if (NOT HW_TOIT_EXE)
```
Comment (kasperl) [inline tests/hw/CMakeLists.txt:22]:
```suggestion
if (NOT TOIT_EXE_HW)
```

File: tests/hw/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  message(FATAL_ERROR "HW_TOIT_EXE not set")
```
Comment (kasperl) [inline tests/hw/CMakeLists.txt:23]:
```suggestion
  message(FATAL_ERROR "TOIT_EXE_HW not set")
```

File: tests/hw/pi/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  )
+
+foreach(file ${TOIT_TESTS})
+  get_filename_component(name ${file} NAME_WE)
+  set(toit_file ${CMAKE_CURRENT_SOURCE_DIR}/${file})
+  set(test_name ${name})
+
+  add_test(
+    NAME ${file}
+    COMMAND ${HW_TOIT_EXE} ${toit_file}
```
Comment (kasperl) [inline tests/hw/pi/CMakeLists.txt:31]:
```suggestion
    COMMAND ${TOIT_EXE_HW} ${toit_file}
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (floitsch) [reply .github/workflows/ci.yml:78]: done.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (floitsch) [reply .github/workflows/ci.yml]: done.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (floitsch) [reply .github/workflows/ci.yml]: done.

File: Makefile (hunk 3 above)
Comment (floitsch) [reply Makefile]: removed the second part of the sentence.
done.

File: Makefile (hunk 6 above)
Comment (floitsch) [reply Makefile]: done.

File: tests/hw/CMakeLists.txt (hunk 7 above)
Comment (floitsch) [reply tests/hw/CMakeLists.txt]: done.

File: tests/hw/CMakeLists.txt (hunk 8 above)
Comment (floitsch) [reply tests/hw/CMakeLists.txt]: done.

File: tests/hw/CMakeLists.txt (hunk 9 above)
Comment (floitsch) [reply tests/hw/CMakeLists.txt]: done.

File: tests/hw/pi/CMakeLists.txt (hunk 10 above)
Comment (floitsch) [reply tests/hw/pi/CMakeLists.txt]: done.

**Pull request merged**